cdap-metrics-sidecar/src/test/java/io/cdap/cdap/jmx/metrics/JMXMetricsCollectorTest.java

```diff
@@ -0,0 +1,136 @@
+import io.cdap.cdap.api.Environment;
+import io.cdap.cdap.api.metrics.MetricsCollectionService;
+import io.cdap.cdap.api.metrics.MetricsContext;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.proto.id.NamespaceId;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.net.MalformedURLException;
+import java.rmi.registry.LocateRegistry;
+import java.util.Map;
+import javax.management.MBeanServer;
+import javax.management.remote.JMXConnectorServer;
+import javax.management.remote.JMXConnectorServerFactory;
+import javax.management.remote.JMXServiceURL;
+
+import static org.mockito.Matchers.anyLong;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class JMXMetricsCollectorTest {
+  private static final int SERVER_PORT = 11023;
+  private static final String SERVICE_NAME = "test-service";
+  private static JMXConnectorServer svr;
+
+  private static final Map<String, String> METRICS_CONTEXT = ImmutableMap.of(
+    Constants.Metrics.Tag.NAMESPACE, NamespaceId.SYSTEM.getNamespace(),
+    Constants.Metrics.Tag.COMPONENT, SERVICE_NAME);
+
+  @Mock
+  private MetricsCollectionService mockMetricsService;
+  @Mock
+  private MetricsContext mockContext;
+  @Mock
+  private Environment mockEnv;
+
+  @Before
+  public void beforeEach() {
+    MockitoAnnotations.initMocks(this);
+    when(mockMetricsService.getContext(METRICS_CONTEXT)).thenReturn(mockContext);
+    when(mockEnv.getVariable("HOSTNAME")).thenReturn("test-host");
+  }
+
+  @BeforeClass
+  public static void setupClass() throws IOException {
+    svr = createJMXConnectorServer(SERVER_PORT);
+    svr.start();
+    Assert.assertEquals(svr.isActive(), true);
+  }
+
+  @AfterClass
+  public static void teardownClass() throws IOException {
+    svr.stop();
+  }
+
+  @Test
+  public void testInvalidPortInConfig() throws InterruptedException, MalformedURLException {
+    when(mockEnv.getVariable("SERVICE_NAME")).thenReturn(SERVICE_NAME);
+    CConfiguration cConf = CConfiguration.create();
+    cConf.setInt(Constants.JMXMetricsCollector.POLL_INTERVAL, 100);
+    cConf.setInt(Constants.JMXMetricsCollector.SERVER_PORT, -1);
+    JMXMetricsCollector jmxMetrics = new JMXMetricsCollector(cConf, mockMetricsService, mockEnv);
+    jmxMetrics.start();
+    Thread.sleep(200);
+    jmxMetrics.stop();
+    verify(mockContext, never()).gauge(anyString(), anyLong());
+  }
+
+  @Test
+  public void testMissingServiceName() throws InterruptedException, MalformedURLException {
```

Comment:
For invalid port and service name, can we throw exception? It is better than silently running the thread and not collecting any metrics.

Reply:
Done